### ISSUE TITLE
`cov_adj()` additional arguments

### ISSUE THREAD

**Issue by benthestatistician** [benbhansen-stats/propertee#4]
`cov_adj()` is intended as a variant of `predict()` that assembles independent variables to predict off of in manner better suited to our purposes, and also sanitizes values of the treatment variable (cf #3). 

1. Should we present it as a wrapper to `predict()`, so that it can be expected to function whenever there's a suitable `predict()` method and it accepts whatever `...` arguments that method does?  Update: :-1: ~~I'd like to offer users a way to set those arguments, but I don't have a specific use case. And we'd routinely be setting `predict()`'s `type=` argument to "response", perhaps confounding user expectations based on use of `predict()`.~~
2. Also, given its function of overwriting values of treatment variable(s), `cov_adj()` should have a ~~(mandatory?) argument with which~~ way to identify any post-treatment variables appearing as predictors in its first argument. (The funding proposal presents sample code in which `cov_adj()` calls only for a fitted model, not also the names of the treatment variable(s).  This works only if the treatment variable doesn't appear among the fitted model's independent variables, as occurs e.g. in Peters-Belson modeling. ) ~~This makes `cov_adj()`'s signature even less like that of `predict()`.~~

**Comment by benthestatistician** [comment]
Getting back to point 2 in the issue statement, offering users as a way to flag post-treatment variables other than the treatment variable itself that are to be overwritten with a reference value before creating predictions.  

My original idea was to have an optional `cov_adj()` argument in which such variables would be named explicitly.  As another alternative, how about exporting something like
```r
post_tx <- function(x) x
``` 
and inviting people to use it in model formulas to flag post treatment variables?  So a typical use would be something like
```r
des <- rct_design(z~block(clinic)+uoa(patientID), data=D)
cmod <- glm(y ~ post_tx(dose) + age + comorbidities, data=C)
tmod <- lmitt(y~., design=des, offset=cov_adj(cmod), data=Q)
```
In this usage, each of the predictions would be made using the smallest recorded value of `dose`.  (At least in the first implementation; we might later consider setting `newdata$dose` to the median or mode of `dose`, or its median or mode among controls.)

**Comment by jwasserman2** [comment]
I think I like the original proposal more, @benthestatistician, because the optional argument could take a list that specifies each post-treatment variable with its reference value; I'm thinking like `tidyr::replace_na()`. Our current behavior for finding the reference value for the treatment variable specifically could be done by default, but if the user specifies the treatment variable in this list then we can use the reference value they provide. The additional (minor) benefit of the original approach is that users have to update less of their existing code (i.e. fitting the covariance model) if we add this argument to `cov_adj()`.

**Comment by josherrickson** [comment]
I agree with Josh, I like the separate argument better. That said, we could let `post_tx()` take in an optional `reference=` argument to handle that issue.

**Comment by benthestatistician** [comment]
I like @jwasserman2's idea also, as well as the `post_tx()` concept w/ @josherrickson's suggestion.  If we go the JW route, we might also consider the UI of [`base::transform()`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/transform) as opposed to `tidyr::replace_na()`'s, adding a ... to `cov_adj()` where users put one or more expressions stating which variables are to be overwritten before creating the predictions. 

1. If we're offering to compute the replacement value for `var2` when the user doesn't care to specify one, we could let them just give `var2` as one of the expressions. But when specifying a named list, one would have to write out `var2=NULL`, which seems verbose and non-intuitive. 
```r
> list(v1=5, v2=NULL)
$v1
[1] 5

$v2
NULL
> list(v1=5, v2)
Error: object 'v2' not found
```
(I suppose this could be mitigated by allowing `list(v1=5, v2="default")`.)
2.  `transform()` offers you more options, you can do stuff like the following.  
```r
> transform(airquality, Ozone=mean(Ozone[Month==5], na.rm=T)) |> head()
  Ozone Solar.R Wind Temp Month Day
1  23.6     190  7.4   67     5   1
2  23.6     118  8.0   72     5   2
3  23.6     149 12.6   74     5   3
4  23.6     313 11.5   62     5   4
5  23.6      NA 14.3   56     5   5
6  23.6      NA 14.9   66     5   6
> airquality |> head()
  Ozone Solar.R Wind Temp Month Day
1    41     190  7.4   67     5   1
2    36     118  8.0   72     5   2
3    12     149 12.6   74     5   3
4    18     313 11.5   62     5   4
5    NA      NA 14.3   56     5   5
6    28      NA 14.9   66     5   6
```
3. What we're doing is more like what `transform()` does than what `replace_na()` does: we're overwriting all values of the variable, not just those with NAs, or those in the treatment group.  So the analogy to `transform()` points the user in the right direction. 

OTOH,  `transform()` comes with a warning label I don't fully understand:
> For programming it is better to use the standard subsetting arithmetic functions, and in particular the non-standard evaluation of argument transform can have unanticipated consequences.

I didn't see anything specific to scare me off in [this discussion](https://forum.posit.co/t/teaching-dplyr-functions-which-have-base-r-equivalents/3775/19) of `dplyr::mutate()` vs `base::transform()`, but I also didn't consider scenarios that might come up in programming (as Wickham discusses [here](https://adv-r.hadley.nz/evaluation.html#base-evaluation) for `base::subset()`). If one of you does understand it and thinks it should be scaring us, do speak up.

**Comment by josherrickson** [comment]
The fact that the example for `transform` in the documentation uses `attach` makes me very wary of it...

Are you suggesting syntax like this?

```r
cmod <- glm(y ~ post_tx(dose) + age + comorbidities, data=C)
tmod <- lmitt(y~., design=des, offset=cov_adj(cmod, age = 45), data=Q)
```

First, how does that solve the problem of users letting us pick the value? Were you thinking `cov_adj(cmod, age = 45, comborbidities)`? I think that would error, you'd need to pass comorbidities as a string. If we're going this approach, I'd suggest we define a function `default`, so then the user could do `cov_adj(cmod, age = 45, comorbidites = default)`. (This is similar to the family argument to `glm` and similar).

Second, I worry about us locking into the `...`. What if down the road we want to use the `...` to pass arguments through `cov_adj` to something lower, e.g. `.make_PreSandwichlayer` which does in fact take in a `...` that we're not currently passing down.

I remain on the side of a list, perhaps with that `default()` function. Something like `cov_adj(cmod, post_tx = list(age = 45, comorbidities = default))`.
